packages/next-server/lib/head.tsx

```diff
@@ -137,32 +137,54 @@ function reduceComponents(
     })
 }
 
-const Effect = withSideEffect()
-
 /**
  * This component injects elements to `<head>` of your page.
  * To avoid duplicated `tags` in `<head>` you can use the `key` property, which will make sure every tag is only rendered once.
  */
-function Head({ children }: { children: React.ReactNode }) {
-  return (
-    <AmpStateContext.Consumer>
-      {ampState => (
-        <HeadManagerContext.Consumer>
-          {updateHead => (
-            <Effect
-              reduceComponentsToState={reduceComponents}
-              handleStateChange={updateHead}
-              inAmpMode={isInAmpMode(ampState)}
-            >
-              {children}
-            </Effect>
-          )}
-        </HeadManagerContext.Consumer>
-      )}
-    </AmpStateContext.Consumer>
-  )
+function Head({ children }: { children: JSX.Element }) {
+  const ampState = React.useContext(AmpStateContext)
+  const updateHead = React.useContext(HeadManagerContext)
+  const instanceRef = React.useRef(children)
+
+  // Update the instanceRef every render
+  instanceRef.current = children
+
+  const inAmpMode = isInAmpMode(ampState)
+
+  const emitUpdate = React.useCallback(() => {
+    headState = reduceHeadInstances([...headInstanceRefs], {
+      inAmpMode,
+    })
+    if (updateHead) {
+      updateHead(headState)
+    }
+  }, [updateHead, reduceHeadInstances, inAmpMode])
+
+  // Since this is a Set(), we can just always add it
+  headInstanceRefs.add(instanceRef)
+
+  if (typeof window === 'undefined') {
+    emitUpdate()
+  }
+
+  // We want to run this after every single render... right?
```

Comment:
This is not the exact equivalent of https://github.com/zeit/next.js/pull/7937/files#diff-d2ba5b59eb8d9e5d8a2ea25c45833eceL50 right 🤔 So in this case it would trigger more than it would before?

Reply:
Since there is no dependency array, this does fire after every render, exactly like `componentDidUpdate`.

Reply:
I also second that this isn't exactly identical -- previously, `componentDidUpdate` would only fire `emitUpdate()`.

The current implementation, following the semantics of `useEffect`, means every component update will cause a cleanup and re-register, meaning an update now invokes:
```js
// old cleanup
headInstanceRefs.delete(instanceRef) // L175
emitUpdate() // L176
// new run
headInstanceRefs.add(instanceRef) // L164
emitUpdate() // L172
```

To get the exact behavior as previously, we'd almost need to split this into two effects:
```js
  // every component update, including mount, triggers emitUpdate()
  React.useEffect(() => {
    emitUpdate()
  })
  // register an on-unmount effect, n.b. `[]` dep array and a noop main function body
  React.useEffect(() => {
     return () => {
      headInstanceRefs.delete(instanceRef)
      emitUpdate()
    }
  }, [])
```

Reply:
Ahhh. Yep, that makes more sense. I've pushed those changes. 👍